pkg/connector/mapping.go

```diff
@@ -0,0 +1,348 @@
+	return bridgev2.ChatMember{
+		EventSender: bridgev2.EventSender{
+			IsFromMe: isFromMe,
+			Sender:   networkid.UserID(user.IDStr),
+		},
+		UserInfo: &bridgev2.UserInfo{
+			Name:   &user.Name,
+			Avatar: MakeAvatar(user.ProfileImageURL),
+		},
+	}
+}
+
+func (tc *TwitterClient) ParticipantToChatMember(participant types.Participant, isFromMe bool) bridgev2.ChatMember {
+	return bridgev2.ChatMember{
+		EventSender: bridgev2.EventSender{
+			IsFromMe: isFromMe,
+			Sender:   networkid.UserID(participant.UserID),
+		},
+		UserInfo: tc.GetUserInfoBridge(participant.UserID),
+	}
+}
+
+func (tc *TwitterClient) GetUserInfoBridge(userId string) *bridgev2.UserInfo {
+	var userinfo *bridgev2.UserInfo
+	if userCacheEntry, ok := tc.userCache[userId]; ok {
+		userinfo = &bridgev2.UserInfo{
+			Name:        ptr.Ptr(tc.connector.Config.FormatDisplayname(userCacheEntry.ScreenName, userCacheEntry.Name)),
+			Avatar:      MakeAvatar(userCacheEntry.ProfileImageURL),
+			Identifiers: []string{fmt.Sprintf("twitter:%s", userCacheEntry.ScreenName)},
+		}
+	}
+	return userinfo
+}
+
+func (tc *TwitterClient) TwitterAttachmentToMatrix(ctx context.Context, portal *bridgev2.Portal, intent bridgev2.MatrixAPI, attachment *types.Attachment) (*bridgev2.ConvertedMessagePart, []int, error) {
+	var attachmentInfo *types.AttachmentInfo
+	var attachmentURL string
+	var mimeType string
+	var indices []int
+	var msgType bridgeEvt.MessageType
+	extraInfo := map[string]any{}
+	if attachment.Photo != nil {
+		// image attachment
+		attachmentInfo = attachment.Photo
+		mimeType = "image/jpeg" // attachment doesn't include this specifically
+		msgType = bridgeEvt.MsgImage
+		attachmentURL = attachmentInfo.MediaURLHTTPS
+		indices = attachmentInfo.Indices
+	} else if attachment.Video != nil || attachment.AnimatedGif != nil {
+		// video attachment
+		if attachment.AnimatedGif != nil {
+			attachmentInfo = attachment.AnimatedGif
+			extraInfo["fi.mau.gif"] = true
+			extraInfo["fi.mau.loop"] = true
+			extraInfo["fi.mau.autoplay"] = true
+			extraInfo["fi.mau.hide_controls"] = true
+			extraInfo["fi.mau.no_audio"] = true
+		} else {
+			attachmentInfo = attachment.Video
+		}
+		mimeType = "video/mp4"
+		msgType = bridgeEvt.MsgVideo
+
+		highestBitRateVariant, err := attachmentInfo.VideoInfo.GetHighestBitrateVariant()
+		if err != nil {
+			return nil, nil, err
+		}
+		attachmentURL = highestBitRateVariant.URL
+		indices = attachmentInfo.Indices
+	}
+
+	if attachmentInfo == nil || attachmentInfo.IDStr == "" {
+		return nil, nil, fmt.Errorf("missing attachment info")
+	}
+
+	attachmentBytes, err := DownloadPlainFile(ctx, tc.client.GetCookieString(), attachmentURL, "twitter attachment")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	content := convertTwitterAttachmentMetadata(attachmentInfo, mimeType, msgType, attachmentBytes)
+	err = uploadMedia(ctx, portal, intent, attachmentBytes, &content)
+	if err != nil {
+		return nil, nil, err
+	}
```

Comment:
I guess Twitter doesn't allow big files, but might still be better to use UploadMediaStream rather than downloading files to memory

Reply:
Twitter **does** allow big files, hundreds of MBs (in videos only). Photos have a limit of around 5MB.
Videos have like an ambiguous 100-200MB limit (not officially stated anywhere but most people using DMs notice this)
Normal posts have a strict 512MB limit

https://developer.x.com/en/docs/x-api/v1/media/upload-media/uploading-media/media-best-practices

I was looking at the dev guide the other day, I'm not sure if we are doing chunked uploads ourselves ( though this is not the point of your comment )